src/reason_parser.mly

```diff
@@ -2797,8 +2777,36 @@ parenthesized_expr:
 
 %inline simple_expr: simple_expr_call { mkexp_app_rev $startpos $endpos $1 };
 
+simple_expr_no_constructor:
+  mark_position_exp(simple_expr_template(simple_expr_no_constructor)) { $1 };
+
+simple_expr_template_constructor:
```

Comment:
What is meant by "template"?

Reply:
We need to use the same definitions in different contexts.
The part that are shared are defined as parametric/higher-order non-terminals. The parameter is used in place of recursive occurrences of the non-terminal. That way the fix point can be tied at the use site.

Thus it is a template of the actual grammar, with the recursion not fixed; the terminology has the same meaning as `templace` in C++.
The parameter plays the same role as a self-type in an OOP language. Actually, OOP inheritance seems appropriate in a grammar specification language.

Reply:
A small illustration with an extensible arithmetic grammar.

You start with:
```
expr: INT | LPAREN expr RPAREN | expr PLUS expr.
```
You want to add a `TIMES` operator:
```
expr2: expr TIMES expr
```
Fine, but the recursion is done on `expr`, so `TIMES` can only be used at the top-level.

Second attempt, factor the recursion out of `expr`:
```
open_expr<Y>: INT | LPAREN Y RPAREN | Y PLUS Y.
```
Now the original expr can simply be recovered with `expr: open_expr<expr>` (a fix point of open_expr).

The extension:
```
expr2: expr2 TIMES expr2 | open_expr<expr2>
```

The scheme composes (up to LR), so you can keep the extension open:
```
open_expr2<Y>: Y TIMES Y | open_expr<Y>
```

With an OO specification, Y would be `'self`.